### PR TITLE
EZP-31080: Created missing search Criterions

### PR DESCRIPTION
**Pull request by webhdx**
| Question           | Answer
| ------------------ | ------------------
| **JIRA issue**     | [EZP-31080](https://jira.ez.no/browse/EZP-31080)
| **Bug/Improvement**| no
| **New feature**    | yes
| **Target version** | `master`
| **BC breaks**      | no
| **Tests pass**     | yes
| **Doc needed**     | yes

# Description
This PR adds new criterions. REST implementation is provided in https://github.com/ezsystems/ezplatform-rest/pull/39

## IsUserBased
Matches content that are also User instances (usually `user` content type but may very on different installations). Supported operators: `EQ`

#### REST example:
`{"IsUserBasedCriterion": false}`

## IsUserEnabled

Matches only enabled/disabled users (via `ezuser` fieldtype attribute). Supported operators: `EQ`

#### REST example:
`{"IsUserEnabledCriterion": true}`

## UserId

Matches only particular User IDs. Supported operators: `EQ`, `IN`

#### REST example:
`{"UserIdCriterion": '10,14'}`

## UserLogin

Matches only particular User logins. Supported operators: `EQ`, `IN`, `LIKE`
 
#### REST example:
* Equal matches: `{"UserLoginCriterion": 'admin,anonymous'}`
* Matched by pattern: `{"UserLoginCriterion": 'adm*'}`

## UserEmail

Matches only particular User e-mails. Supported operators: `EQ`, `IN`, `LIKE`
 
#### REST example:
* Equal matches: `{"UserEmailCriterion": 'nospam@ez.no'}`
* Matched by pattern: `{"UserEmailCriterion": 'nospam*'}`

## SectionIdentifier

Matches only content in particular sections. Supported operators: `EQ`, `IN`
 
#### REST example:
`{"SectionIdentifierCriterion": 'users'}`

## ObjectStateIdentifier

Matches only content in particular object state. It supports passing object state group as a `$target`. Supported operators: `EQ`, `IN`
 
#### REST example:
`{"ObjectStateIdentifierCriterion": {"value": 'not_locked', "target": 'ez_lock'}}`

**TODO**:
- [x] Implement feature / fix a bug.
- [x] Implement tests.
- [x] Fix new code according to Coding Standards (`$ composer fix-cs`).
- [x] Ask for Code Review.

### CODE REVIEW

File: eZ/Publish/API/Repository/Values/Content/Query/Criterion/UserBased.php [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion\Operator\Specifications;
+
+class UserBased extends Criterion
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/Content/Query/Criterion/UserBased.php]: Please keep the existing convention for the boolean criterion e.g. IsMainLocation, IsFieldEmpty etc.

```suggestion
class IsUserBased extends Criterion
```

File: eZ/Publish/API/Repository/Values/Content/Query/Criterion/UserEnabled.php [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
+namespace eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion;
+use eZ\Publish\API\Repository\Values\Content\Query\Criterion\Operator\Specifications;
+
+class UserEnabled extends Criterion
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/Content/Query/Criterion/UserEnabled.php]: Please keep the existing convention for the boolean criterion e.g. IsMainLocation, IsFieldEmpty etc.. 

```suggestion
class IsUserEnabled extends Criterion
```

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/UserEmail.php [hunk 3]
```diff
@@ -0,0 +1,81 @@
...
+            $expression = $query->expr->like(
+                $this->dbHandler->quoteColumn('email', 't1'),
+                $query->bindValue(
+                    str_replace(
+                        '*',
+                        '%',
+                        addcslashes(
+                            $this->transformationProcessor->transformByGroup(
+                                $criterion->value,
+                                'lowercase'),
```
Comment (adamwojs) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/UserEmail.php]:
```suggestion
                                'lowercase'
                            ),
```

File: eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/UserLogin.php [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+            $expression = $query->expr->like(
+                $this->dbHandler->quoteColumn('login', 't1'),
+                $query->bindValue(
+                    str_replace(
+                        '*',
+                        '%',
+                        addcslashes(
+                            $this->transformationProcessor->transformByGroup(
+                                $criterion->value,
+                                'lowercase'),
```
Comment (adamwojs) [inline eZ/Publish/Core/Search/Legacy/Content/Common/Gateway/CriterionHandler/UserLogin.php]:
```suggestion
                                'lowercase'
                            ),
```

File: eZ/Publish/API/Repository/Values/Content/Query/Criterion/ObjectStateIdentifier.php [hunk 5]
```diff
@@ -0,0 +1,39 @@
+<?php
+declare(strict_types=1);
+
+/**
+ * @copyright Copyright (C) eZ Systems AS. All rights reserved.
+ * @license For full copyright and license information view LICENSE file distributed with this source code.
+ */
```
Comment (adamwojs) [inline eZ/Publish/API/Repository/Values/Content/Query/Criterion/ObjectStateIdentifier.php]: The license header is in the wrong place (also in all other new files).

File: eZ/Publish/API/Repository/Tests/SearchServiceTest.php [hunk 6]
```diff
@@ -656,6 +657,134 @@ public function getContentQuerySearches()
...
+                [
+                    'query' => new Criterion\SectionIdentifier(['users']),
+                    'sortClauses' => [new SortClause\ContentId()],
+                    'limit' => 50,
+                ],
+                $fixtureDir . 'SectionIdentifier.php',
+            ],
+            [
+                [
+                    'query' => new Criterion\ObjectStateIdentifier('not_locked'),
```
Comment (alongosz) [inline eZ/Publish/API/Repository/Tests/SearchServiceTest.php]: I think we already discussed internally also this issue. If you have an object state which shares the same identifier across multiple groups, then this is not very useful. The case might not be as uncommon as you might think.

**Comment by webhdx** [comment]
@alongosz You are a bit unfair here as we spoke previously you were fully aware of this PR using deprecated DatabaseHandler and you told me if this gets merged first you are going to refactor it. I need your guidance then as I'm not too familiar with all needed changes.

File: eZ/Publish/API/Repository/Tests/SearchServiceTest.php (hunk 6 above)
Comment (webhdx) [reply eZ/Publish/API/Repository/Tests/SearchServiceTest.php]: This has been fixed shortly after the demo when you DMed me. It is that the tests are not relying on it. 

ref.  https://github.com/ezsystems/ezpublish-kernel/pull/2952/files#diff-b300f6e65c848771c4916553315ecc83R65

Comment (gggeek) [reply eZ/Publish/API/Repository/Tests/SearchServiceTest.php]: fwiw, in Mig Bundle we do allow usage of state identifier, as it is pretty nice to be able to skip the group identifier. Then the code checks: if the identifier is unique, all god. If it is reused across groups, it throws.

**Comment by alongosz** [comment]
> You are a bit unfair here as we spoke previously you were fully aware of this PR using deprecated DatabaseHandler and you told me if this gets merged first you are going to refactor it. I need your guidance then as I'm not too familiar with all needed changes.

That's true. Moreover I created this mess. I just saw very straightforward possibility of a work-around and wished we discussed this prior implementing (when we spoke, it was after the fact).
I'm gonna handle this as agreed then. 

So the only thing remaining is a test which didn't cover the change - so I thought it wasn't applied.

**Pull request merged**